v3/src/models/data/formula-fn-registry.ts

```diff
@@ -0,0 +1,157 @@
+  return node.compile().evaluate(scope)
+}
+
+export const fnRegistry = {
+  // equal(a, b) or a == b
+  // Note that we need to overwrite default MathJs implementation so we can compare strings like "ABC" == "CDE".
+  // MathJs doesn't allow that by default, as it assumes that equal operator can be used only with numbers.
+  equal: {
+    rawArgs: false,
+    isAggregate: false,
+    evaluate: (a: any, b: any) => {
+      if (Array.isArray(a) && Array.isArray(b)) {
+        return a.map((v, i) => v === b[i])
+      }
+      if (Array.isArray(a) && !Array.isArray(b)) {
+        return a.map((v) => v === b)
+      }
+      if (!Array.isArray(a) && Array.isArray(b)) {
+        return b.map((v) => v === a)
+      }
+      return a === b
+    }
+  },
+
+  // lookupByIndex("dataSetName", "attributeName", index)
+  lookupByIndex: {
+    rawArgs: true,
+    isAggregate: false,
+    validateArguments: (args: MathNode[]): [ConstantNode<string>, ConstantNode<string>, MathNode] => {
+      if (args.length !== 3) {
+        throw new Error(`lookupByIndex function expects exactly 3 arguments, but it received ${args.length}`)
+      }
+      if (!isConstantStringNode(args[0]) || !isConstantStringNode(args[1])) {
+        throw new Error("lookupByIndex function expects first two arguments to be strings " +
+          "and the third one to be a number")
+      }
+      return [args[0], args[1], args[2]]
+    },
+    getDependency: (args: MathNode[]): ILookupDependency => {
+      const validArgs = fnRegistry.lookupByIndex.validateArguments(args)
+      return {
+        type: "lookup",
+        dataSetId: validArgs[0].value,
+        attrId: validArgs[1].value,
+      }
+    },
+    canonicalize: (args: MathNode[], displayNameMap: DisplayNameMap) => {
+      const validArgs = fnRegistry.lookupByIndex.validateArguments(args)
+      const dataSetName = validArgs[0].value
+      const attrName = validArgs[1].value
+      validArgs[0].value = displayNameMap.dataSet[dataSetName]?.id
+      validArgs[1].value = displayNameMap.dataSet[dataSetName]?.attribute[attrName]
+    },
+    evaluate: (args: MathNode[], mathjs: any, scope: FormulaMathJsScope) => {
+      const dataSetId = evaluateNode(args[0], scope)
+      const attrId = evaluateNode(args[1], scope)
+      const zeroBasedIndex = evaluateNode(args[2], scope) - 1
+      return scope.getDataSet(dataSetId)?.getValueAtIndex(zeroBasedIndex, attrId) || ""
+    }
+  },
+
+  // lookupByKey("dataSetName", "attributeName", "keyAttributeName", "keyAttributeValue" | localKeyAttribute)
+  lookupByKey: {
+    rawArgs: true,
+    isAggregate: false,
+    validateArguments: (args: MathNode[]):
+      [ConstantNode<string>, ConstantNode<string>, ConstantNode<string>, MathNode] => {
+      if (args.length !== 4) {
+        throw new Error("lookupByKey function expects exactly 4 arguments")
+      }
+      if (!isConstantStringNode(args[0]) || !isConstantStringNode(args[1]) || !isConstantStringNode(args[2])) {
```

Comment:
Could also use:
```suggestion
      if (!args.slice(0, 3).every(arg => isConstantStringNode(arg)) {
```

Reply:
It seems that it won't work in this particular case. These simple `if` guards allow TS to assume that in the following lines args[0], args[1], and args[1] are instances of `ConstantNode<string>`. Your expression is nicer and more concise, but TS is not smart enough to parse it and infer correct types from it. It'll generate an error in the next line:
` Type 'MathNode' is not assignable to type 'ConstantNode<string>'.`